### PR TITLE
Add DynamicFilter to ConnectorRecordSetProvider SPI

### PR DESCRIPTION
**Pull request by jerryleooo**
https://github.com/trinodb/trino/issues/8045

### CODE REVIEW

**Comment by jerryleooo** [comment]
reopen to refresh the CLA

**Comment by martint** [comment]
@cla-bot check

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java [hunk 1]
```diff
@@ -13,10 +13,36 @@
...
+        return getRecordSet(transaction, session, split, table, columns, dynamicFilter.getCurrentPredicate());
+    }
+
+    default RecordSet getRecordSet(
+            ConnectorTransactionHandle transaction,
+            ConnectorSession session,
+            ConnectorSplit split,
+            ConnectorTableHandle table,
+            List<? extends ColumnHandle> columns,
+            TupleDomain<ColumnHandle> dynamicFilter)
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java:40]: This `TupleDomain` variant in `ConnectorPageSourceProvider` is deprecated (it pre-dates `DynamicFilter`). We should only have a single method `DynamicFilter`, which contains `DynamicFilter`.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java [hunk 2]
```diff
@@ -13,10 +13,36 @@
...
+    {
+        // By default, ignore dynamic filter (as it is an optimization and doesn't affect correctness).
+        return getRecordSet(transaction, session, split, table, columns);
+    }
+
     default RecordSet getRecordSet(
```
Comment (electrum) [inline core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java:46]: Please mark this as `@Deprecated`. Long term, we should only have the `DynamicFilter` version of this method, and remove the others.

**Review by electrum — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java (hunk 1 above)
Comment (jerryleooo) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java:40]: Hi @electrum thanks for the review. The reason I added it here is https://github.com/trinodb/trino/blob/fe608f2723842037ff620d612a706900e79c52c8/core/trino-main/src/main/java/io/trino/split/RecordPageSourceProvider.java#L48 
Maybe I should change all places?

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java (hunk 2 above)
Comment (jerryleooo) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java:46]: Sure, will do.

File: core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java (hunk 1 above)
Comment (raunaqmorarka) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java:40]: @jerryleooo i got rid of the deprecated APIs in ConnectorPageSourceProvider in https://github.com/trinodb/trino/pull/8077 to avoid confusion around this

Comment (jerryleooo) [reply core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorRecordSetProvider.java:40]: Thanks, @raunaqmorarka , then I may just wait for a while. If #8077 needs more time to merge, I will just leave a TODO here.

**Review by raunaqmorarka — DISMISSED** [review]
I would reword the commit message to :
Add DynamicFilter to ConnectorRecordSetProvider SPI

**Comment by jerryleooo** [comment]
@raunaqmorarka updated. Then how to ask for a merge? Thanks

**Comment by raunaqmorarka** [comment]
> @raunaqmorarka updated. Then how to ask for a merge? Thanks

@jerryleooo we need to update ClassLoaderSafeConnectorRecordSetProvider as well

**Comment by jerryleooo** [comment]
@electrum @raunaqmorarka all checks passed, need I squash the commits into one? Thanks

**Comment by jerryleooo** [comment]
@electrum can I get this merged?

**Comment by raunaqmorarka** [comment]
If this is really needed let's do this as a commit in the PR which adds DF for JDBC sources.
It's probably better to use JdbcSplitManager instead per https://github.com/trinodb/trino/pull/8137#discussion_r908068229
